### PR TITLE
Skip incompatible dtypes in op_test using OpSchema | test(torchlib)

### PR DESCRIPTION
**Pull request by justinchuby**
This requires ONNX 1.14. I am skipping the op test for ONNX<1.14 altogether.

I tested locally to make sure no float32 test is skipped with this mechanism, and that it skips properly according to the type annotation.

Currently no new dtypes are enabled because there are too many failures.

Requires 
- #701 
- #698

### CODE REVIEW

File: onnxscript/tests/function_libs/torch_lib/ops_test_common.py [hunk 1]
```diff
@@ -355,6 +355,41 @@ def _format_model_and_input_information(onnx_model, inputs):
...
+}
+
+
+def dtype_op_schema_compatible(dtype: torch.dtype, schema: onnx.defs.OpSchema) -> bool:
+    if not schema.inputs:
+        # If there are no inputs, we can't check compatibility. Assume it is compatible.
+        # e.g. aten_randn has only attributes.
+        return True
+    if schema.inputs[0].name not in {"self", "input"}:
+        # If the first input is not self or input, it is usually another input that is not
```
Comment (BowenBao) [inline onnxscript/tests/function_libs/torch_lib/ops_test_common.py:380]: I'm not following this comment section... I don't get the example too.

File: onnxscript/tests/function_libs/torch_lib/ops_test_common.py [hunk 2]
```diff
@@ -355,6 +355,41 @@ def _format_model_and_input_information(onnx_model, inputs):
     )
 
 
+TORCH_DTYPE_TO_ONNX_STRING = {
+    torch.bool: "tensor(bool)",
```
Comment (BowenBao) [inline onnxscript/tests/function_libs/torch_lib/ops_test_common.py:359]: nit: Commented elsewhere before but don't recall precisely. Does it has to be plain string for onnx types?

File: onnxscript/tests/function_libs/torch_lib/ops_test_common.py [hunk 3]
```diff
@@ -355,6 +355,41 @@ def _format_model_and_input_information(onnx_model, inputs):
...
+    torch.float16: "tensor(float16)",
+    torch.float32: "tensor(float)",
+    torch.float64: "tensor(double)",
+    torch.complex64: "tensor(complex64)",
+    torch.complex128: "tensor(complex128)",
+    torch.bfloat16: "tensor(bfloat16)",
+}
+
+
+def dtype_op_schema_compatible(dtype: torch.dtype, schema: onnx.defs.OpSchema) -> bool:
```
Comment (BowenBao) [inline onnxscript/tests/function_libs/torch_lib/ops_test_common.py:374]: I think it's worth commenting to explain the context for `dtype`. It is the "dtype setting" introduced by opinfo tests, so not any average input dtype. Feels like it holds the assumption that to some extent, all arguments outputs should have the same `dtype`.

I think it helps also to brief what it means if a schema is compatible to a dtype. Do all arguments, outputs have to share that same type?

File: onnxscript/tests/function_libs/torch_lib/ops_test_common.py [hunk 4]
```diff
@@ -355,6 +355,41 @@ def _format_model_and_input_information(onnx_model, inputs):
...
+    if not schema.inputs:
+        # If there are no inputs, we can't check compatibility. Assume it is compatible.
+        # e.g. aten_randn has only attributes.
+        return True
+    if schema.inputs[0].name not in {"self", "input"}:
+        # If the first input is not self or input, it is usually another input that is not
+        # the same type as the output. Assume it is compatible.
+        # e.g. aten_add has inputs "self" and "other".
+        return True
+    first_input_dtype = schema.inputs[0].type_str
```
Comment (BowenBao) [inline onnxscript/tests/function_libs/torch_lib/ops_test_common.py:384]: Related to the comment on this function, I'm not quite following the logic of below lines. Could you briefly explain?

Comment (BowenBao) [reply onnxscript/tests/function_libs/torch_lib/ops_test_common.py:384]: Thoughts are clearer after realizing `first_input_dtype` is a type param str like `TReal`. Is it guaranteed that `schema.inputs[0].type_str` is always a type param str?

File: onnxscript/tests/function_libs/torch_lib/ops_test_common.py (hunk 2 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/ops_test_common.py:359]: Yes, it needs to be a plain string in onnx's type constraints

Comment (BowenBao) [reply onnxscript/tests/function_libs/torch_lib/ops_test_common.py:359]: I was wondering if there is anything like `str(onnx.BoolTensorType)`/`onnx.BoolTensorType.to_string()`

File: onnxscript/tests/function_libs/torch_lib/ops_test_common.py (hunk 3 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/ops_test_common.py:374]: Added When a dtype is "compatible" with the schema, it means we can use the dtype to create sample inputs by OpInfo to test the ONNX function and expect outputs to match.

File: onnxscript/tests/function_libs/torch_lib/ops_test_common.py (hunk 4 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/ops_test_common.py:400]: > Is it guaranteed that schema.inputs[0].type_str is always a type param str?

Do you mean it can be `sequence(tensor(int64))`? I realized that's a case I didn't handle.

Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/ops_test_common.py]: But it's also just ops like `def aten_stack(tensors: Sequence[TTensorOrString], dim: int = 0)` whose first input is name "self" or "input" so we always test it on all dtypes.

Comment (BowenBao) [reply onnxscript/tests/function_libs/torch_lib/ops_test_common.py]: Cool for finding an edge case, but to your question slightly no, I was just wondering if it is legal to set `tensor(float32)` for `schema.inputs[0].type_str`.  I think it shouldn't, otherwise it becomes weird for `schema.type_constraints`.

Comment (BowenBao) [reply onnxscript/tests/function_libs/torch_lib/ops_test_common.py]: Or maybe it's legal when that's the only type supported, and type_constraints could be empty?

File: onnxscript/tests/function_libs/torch_lib/ops_test_common.py (hunk 3 above)
Comment (BowenBao) [reply onnxscript/tests/function_libs/torch_lib/ops_test_common.py:374]: I'm very curious if there are edge cases when sample inputs are different type tensors and how opinfo handles it. Thanks for comments. They are very clear now.

File: onnxscript/tests/function_libs/torch_lib/ops_test_common.py (hunk 4 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/ops_test_common.py]: type_str will be a name (`T1`, `TFloat` etc.) We use this name to find the allowed types in type_constraints

File: onnxscript/tests/function_libs/torch_lib/ops_test.py [hunk 5]
```diff
@@ -162,6 +162,11 @@ def run_test_output_match(
     # An example is nn.functional.upsample_nearest2d, which has a different signature
     # than the aten operator upsample_nearest2d
     onnx_function, input_wrangler = _split_function_and_wrangler(onnx_function_and_wrangler)
+    if not ops_test_common.dtype_op_schema_compatible(dtype, onnx_function.op_schema):
+        test_suite.skipTest(
+            f"dtype '{dtype}' is not supported by the op '{op.name}'. "
+            f"Type constraints: {onnx_function.op_schema.type_constraints}"
+        )
```
Comment (titaiwangms) [inline onnxscript/tests/function_libs/torch_lib/ops_test.py:169]: I have a concern on skipping tests with the manual annotated type constraints. If we wrongly annotate an Op inputs, it could skip the tests it shouldn't, which makes us miss the very first opportunity to catch it.

File: onnxscript/tests/function_libs/torch_lib/ops_test_common.py (hunk 4 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/ops_test_common.py]: That's guaranteed to not be `tensor(float32)` if that was the question

File: onnxscript/tests/function_libs/torch_lib/ops_test_common.py [hunk 6]
```diff
@@ -355,6 +355,58 @@ def _format_model_and_input_information(onnx_model, inputs):
...
+    torch.float16: "tensor(float16)",
+    torch.float32: "tensor(float)",
+    torch.float64: "tensor(double)",
+    torch.complex64: "tensor(complex64)",
+    torch.complex128: "tensor(complex128)",
+    torch.bfloat16: "tensor(bfloat16)",
+}
+
+
+def dtype_op_schema_compatible(dtype: torch.dtype, schema: onnx.defs.OpSchema) -> bool:
```
Comment (BowenBao) [inline onnxscript/tests/function_libs/torch_lib/ops_test_common.py:374]: I think this is good for a quick and general guard for tests and evolve from there.

What are your thoughts on more fine-grained type checking that validates type of each concrete input to the type of the matching input in schema?

**Review by BowenBao — APPROVED** [review]

File: onnxscript/tests/function_libs/torch_lib/ops_test_common.py (hunk 4 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/ops_test_common.py]: > if it is legal to set `tensor(float32)` for `schema.inputs[0].type_str`. 

It is not. Or I should say we _can_ (we may be able to use it as a name), but it still needs to correspond to an entry in type_constraints. Otherwise there will an initialization error in OpSchema I believe.

File: onnxscript/tests/function_libs/torch_lib/ops_test.py (hunk 5 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/ops_test.py:169]: That's true. Any suggestions?

Comment (BowenBao) [reply onnxscript/tests/function_libs/torch_lib/ops_test.py:169]: wonder if we can xfail

File: onnxscript/tests/function_libs/torch_lib/ops_test_common.py (hunk 6 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/ops_test_common.py:374]: I think we will need to build the same logic in the dispatcher. We can evolve that can use it here. For now I think it's good to use logic that is easy to understand and get right for running test cases. It is always better to say inputs are compatible than not so we don't miss test cases. We can always put an xfail on them.

Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/ops_test_common.py:374]: Also this logic is for reducing the amount of tests we need to manually skip, to accuracy isn't super important here (as long as we don't skip more than we should)

File: onnxscript/tests/function_libs/torch_lib/ops_test.py (hunk 5 above)
Comment (titaiwangms) [reply onnxscript/tests/function_libs/torch_lib/ops_test.py:169]: Based on what I found in https://github.com/pytorch/pytorch/pull/100265, indeed, if the type constraints are from onnx spec, most likely it's xfail-lable.

File: onnxscript/tests/function_libs/torch_lib/ops_test_common.py (hunk 2 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/ops_test_common.py:359]: I see. There probably is but we still need a mapping from torch types to onnx types. Any suggestions?

File: onnxscript/tests/function_libs/torch_lib/ops_test_common.py (hunk 1 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/ops_test_common.py]: I think this is resolved below.

File: onnxscript/tests/function_libs/torch_lib/ops_test.py (hunk 5 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/ops_test.py:169]: Sometimes it will xpass if we do xfail because ort _can_ choose to implement the op for some types intentionally or unintentionally. So that's a little tricky

Comment (titaiwangms) [reply onnxscript/tests/function_libs/torch_lib/ops_test.py:169]: How about put the logic after runs? Just a thought, if it fails, then we check the dtype is supported or not.

Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/ops_test.py:169]: Hmm but that’s still going to skip the tests where the annotation is incorrect?

Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/ops_test.py:169]: I thought about this again. Maybe that’s ok for now since if we declare that something is not supported via the annotations, then the exporter will just need to bridge the types via type promotion logic. When we see unnecessary type promotion in the graph we can then come back and make changes to the annotations.

Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/ops_test.py:169]: Merging for now. Feel free to share more thoughts!
